File: backend/api/routers/preferences.py
from datetime import date

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from supabase import Client

from api.dependencies import get_supabase

router = APIRouter(prefix="/preferences", tags=["preferences"])


class PreferencesTravel(BaseModel):
    member_id: str
    trip_id: str
    origin_city: str
    origin_airport: str
    departure_date: date
    return_date: date
    flight_budget: int
    departure_time_pref: str
    date_flexibility: str


class PreferencesStay(BaseModel):
    member_id: str
    trip_id: str
    budget_min: int
    budget_max: int
    property_types: list[str]
    vibes: list[str]
    needs: list[str]


class PreferencesFood(BaseModel):
    member_id: str
    trip_id: str
    cuisines: list[str]
    dietary: list[str]
    meal_budget: str


class PreferencesActivities(BaseModel):
    member_id: str
    trip_id: str
    interests: list[str]
    pace: str
    must_sees: str


def upsert_preferences(supabase: Client, table: str, payload: dict) -> dict:
    try:
        result = supabase.table(table).upsert(
            payload, on_conflict="member_id,trip_id"
        ).execute()
    except Exception as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    if not result.data:
        raise HTTPException(status_code=500, detail=f"Failed to save {table}")
    return result.data[0]


def fetch_member_preferences(supabase: Client, trip_id: str, member_id: str) -> dict:
    def fetch_one(table: str):
        try:
            result = (
                supabase.table(table).select("*").eq("trip_id", trip_id)
                .eq("member_id", member_id).execute()
            )
        except Exception:
            return None
        return result.data[0] if result.data else None

    return {
        "travel": fetch_one("preferences_travel"),
        "stay": fetch_one("preferences_stay"),
        "food": fetch_one("preferences_food"),
        "activities": fetch_one("preferences_activities"),
    }


@router.post("/travel")
def save_travel_preferences(
    preferences: PreferencesTravel,
    supabase: Client = Depends(get_supabase),
):
    return upsert_preferences(
        supabase, "preferences_travel", preferences.model_dump(mode="json")
    )


@router.post("/stay")
def save_stay_preferences(
    preferences: PreferencesStay,
    supabase: Client = Depends(get_supabase),
):
    return upsert_preferences(
        supabase, "preferences_stay", preferences.model_dump(mode="json")
    )


@router.post("/food")
def save_food_preferences(
    preferences: PreferencesFood,
    supabase: Client = Depends(get_supabase),
):
    return upsert_preferences(
        supabase, "preferences_food", preferences.model_dump(mode="json")
    )


@router.post("/activities")
def save_activities_preferences(
    preferences: PreferencesActivities,
    supabase: Client = Depends(get_supabase),
):
    return upsert_preferences(
        supabase, "preferences_activities", preferences.model_dump(mode="json")
    )


@router.get("/{trip_id}/{member_id}")
def get_member_preferences(
    trip_id: str,
    member_id: str,
    supabase: Client = Depends(get_supabase),
):
    return fetch_member_preferences(supabase, trip_id, member_id)

File: lib/server/preferences.ts
import { API_BASE } from "@/lib/server/api-base";

/**
 * Server-only read of a member's saved preferences across all four
 * categories. Powers the wizard's pre-fill and the trip page's completion
 * checkmarks — never imported from a "use client" file.
 */

export interface TravelPreferences {
  id: string;
  member_id: string;
  trip_id: string;
  origin_city: string;
  origin_airport: string;
  departure_date: string;
  return_date: string;
  flight_budget: number;
  departure_time_pref: string;
  date_flexibility: string;
}

export interface StayPreferences {
  id: string;
  member_id: string;
  trip_id: string;
  budget_min: number;
  budget_max: number;
  property_types: string[];
  vibes: string[];
  needs: string[];
}

export interface FoodPreferences {
  id: string;
  member_id: string;
  trip_id: string;
  cuisines: string[];
  dietary: string[];
  meal_budget: string;
}

export interface ActivitiesPreferences {
  id: string;
  member_id: string;
  trip_id: string;
  interests: string[];
  pace: string;
  must_sees: string;
}

export interface MemberPreferences {
  travel: TravelPreferences | null;
  stay: StayPreferences | null;
  food: FoodPreferences | null;
  activities: ActivitiesPreferences | null;
}

const EMPTY_PREFERENCES: MemberPreferences = {
  travel: null,
  stay: null,
  food: null,
  activities: null,
};

/** §12: an unreachable preferences service degrades to "nothing filled in yet". */
export async function getMemberPreferences(
  tripId: string,
  memberId: string,
): Promise<MemberPreferences> {
  try {
    const url = new URL(`/preferences/${tripId}/${memberId}`, API_BASE);
    const response = await fetch(url, { cache: "no-store" });
    if (!response.ok) return EMPTY_PREFERENCES;

    return (await response.json()) as MemberPreferences;
  } catch {
    return EMPTY_PREFERENCES;
  }
}
